internal/gps/source_manager.go

```diff
@@ -578,6 +578,18 @@ func (sm *SourceMgr) InferConstraint(s string, pi ProjectIdentifier) (Constraint
 		return version.Unpair(), nil
 	}
 
+	// Abbreviated git revision? If the string is present, there's a good shot of
+	// this.
+	if present, _ := sm.RevisionPresentIn(pi, Revision(s)); present {
+		srcg, err := sm.srcCoord.getSourceGatewayFor(context.TODO(), pi)
```

Comment:
My concern here was just in adding another public method to the SourceManager interface, which is already very large. What would you think of a private method on the *sourceManager here to encapsulate this behavior?

`(*sourceManager).RevisionPresentIn` already digs into the source gateway in its implementation, so this would be similar, in my head.